### PR TITLE
Rostislav / 73993 / Task - TS migration of ProfitTable

### PR DESCRIPTION
**Pull request by rostislav-deriv**
## Changes:

Please include a summary of the change and which issue is fixed below:
-   ...

## When you need to add unit test

-   [ ] If this change disrupt current flow
-   [ ] If this change is adding new flow

## When you need to add integration test

-   [ ] If components from external libraries are being used to define the flow, e.g. @deriv/components
-   [ ] If it relies on a very specific set of props with no default behavior for the current component.

## Test coverage checklist (for reviewer)

-   [ ] Ensure utility / function has a test case 
-   [ ] Ensure all the tests are passing

## Type of change

-   [ ] Bug fix
-   [ ] New feature
-   [ ] Update feature
-   [ ] Refactor code
-   [ ] Translation to code
-   [ ] Translation to crowdin
-   [ ] Script configuration
-   [ ] Improve performance
-   [ ] Style only
-   [ ] Dependency update
-   [ ] Documentation update
-   [ ] Release

### CODE REVIEW

File: packages/reports/src/Components/Form/CompositeCalendar/composite-calendar-mobile.tsx [hunk 1]
```diff
@@ -42,7 +42,7 @@ export const RadioButton = ({ id, className, selected_value, value, label, onCha
 };
 const CUSTOM_KEY = 'custom';
 
-type TInputDateRange = {
+export type TInputDateRange = {
```
Comment (mahdiyeh-deriv) [inline packages/reports/src/Components/Form/CompositeCalendar/composite-calendar-mobile.tsx]: please move reusable types to the `packages/reports/src/Types/common-prop.type.ts`

File: packages/reports/src/Constants/data-table-constants.tsx [hunk 2]
```diff
@@ -48,10 +48,16 @@ type THeaderProps = {
     is_vanilla: boolean;
 };
 
-type TColumnTemplateType = {
+export type TColIndex =
```
Comment (mahdiyeh-deriv) [inline packages/reports/src/Constants/data-table-constants.tsx]: same here

File: packages/reports/src/Constants/data-table-constants.tsx [hunk 3]
```diff
@@ -48,10 +48,16 @@ type THeaderProps = {
...
+    | ReturnType<typeof getStatementTableColumnsTemplate>[number]['col_index']
+    | ReturnType<typeof getProfitTableColumnsTemplate>[number]['col_index']
+    | ReturnType<typeof getOpenPositionsColumnsTemplate>[number]['col_index']
+    | ReturnType<typeof getMultiplierOpenPositionsColumnsTemplate>[number]['col_index'];
+
+export type TColumnTemplateType = {
```
Comment (mahdiyeh-deriv) [inline packages/reports/src/Constants/data-table-constants.tsx]: same

File: packages/reports/src/Containers/profit-table.tsx [hunk 4]
```diff
@@ -1,31 +1,59 @@
...
-import { PropTypes as MobxPropTypes } from 'mobx-react';
 import React from 'react';
 import { withRouter } from 'react-router';
 import { DesktopWrapper, MobileWrapper, DataList, DataTable } from '@deriv/components';
-import { extractInfoFromShortcode, isForwardStarting, getUnsupportedContracts, getContractPath } from '@deriv/shared';
+import {
+    extractInfoFromShortcode,
+    isForwardStarting,
+    getContractPath,
+    getSupportedContracts,
```
Comment (mahdiyeh-deriv) [inline packages/reports/src/Containers/profit-table.tsx:9]: nice :+1: 
please change the import path for this utility in other places of `reports` package
for example: `packages/reports/src/Containers/statement.tsx `

Comment (mahdiyeh-deriv) [reply packages/reports/src/Containers/profit-table.tsx:9]: please check the other utilities of `_common/contract` file and if the utility is available in the shared package, remove it from `_common/contract`

File: packages/reports/src/Containers/profit-table.tsx [hunk 5]
```diff
@@ -1,31 +1,59 @@
...
 import CompositeCalendar from '../Components/Form/CompositeCalendar';
-import { getSupportedContracts } from '_common/contract';
+import { TInputDateRange } from '../Components/Form/CompositeCalendar/composite-calendar-mobile';
 
 import { connect } from 'Stores/connect';
 import EmptyTradeHistoryMessage from '../Components/empty-trade-history-message';
 import PlaceholderComponent from '../Components/placeholder-component';
 import { ReportsMeta } from '../Components/reports-meta';
-import { getProfitTableColumnsTemplate } from 'Constants/data-table-constants';
+import { getProfitTableColumnsTemplate, TColIndex, TColumnTemplateType } from 'Constants/data-table-constants';
```
Comment (mahdiyeh-deriv) [inline packages/reports/src/Containers/profit-table.tsx]: move all of these types to the `packages/reports/src/Types/common-prop.type.ts`

File: packages/reports/src/_common/contract.tsx [hunk 6]
```diff
@@ -265,25 +259,24 @@ export const getSupportedContracts = is_high_low => ({
...
-        position: 'top',
-    },
-    VANILLALONGPUT: {
-        name: <Localize i18n_default_text='Put' />,
-        position: 'bottom',
-    },
 });
 
-export const getContractConfig = is_high_low => ({
+export type TContractType =
```
Comment (mahdiyeh-deriv) [inline packages/reports/src/_common/contract.tsx]: same here

File: packages/reports/src/Components/Form/CompositeCalendar/composite-calendar-mobile.tsx [hunk 7]
```diff
@@ -3,6 +3,7 @@ import React from 'react';
 import { Button, DatePicker, Icon, InputField, MobileDialog, Text } from '@deriv/components';
 import { localize } from '@deriv/translations';
 import { toMoment } from '@deriv/shared';
+import { TInputDateRange } from '../../../Types';
```
Comment (mahdiyeh-deriv) [inline packages/reports/src/Components/Form/CompositeCalendar/composite-calendar-mobile.tsx]: Please add `Types` alias in the `packages/reports/build/constants.js` file so you could get benefit of using absolute path
```suggestion
import { TInputDateRange } from 'Types';
```

File: packages/reports/src/Constants/data-table-constants.tsx [hunk 8]
```diff
@@ -7,15 +7,15 @@ import {
     getTotalProfit,
     shouldShowCancellation,
     getGrowthRatePercentage,
+    getCardLabels,
 } from '@deriv/shared';
 import { localize, Localize } from '@deriv/translations';
 import ProgressSliderStream from '../Containers/progress-slider-stream';
-
-import { getCardLabels } from '_common/contract';
+import { TCellContentProps, THeaderProps } from '../Types';
```
Comment (mahdiyeh-deriv) [inline packages/reports/src/Constants/data-table-constants.tsx]: same

File: packages/reports/src/Containers/profit-table.tsx [hunk 9]
```diff
@@ -1,31 +1,58 @@
...
-import { getSupportedContracts } from '_common/contract';
+import { TInputDateRange, TColIndex, TColumnTemplateType, TContractType } from 'Types';
 
 import { connect } from 'Stores/connect';
 import EmptyTradeHistoryMessage from '../Components/empty-trade-history-message';
 import PlaceholderComponent from '../Components/placeholder-component';
 import { ReportsMeta } from '../Components/reports-meta';
 import { getProfitTableColumnsTemplate } from 'Constants/data-table-constants';
+import { TRootStore } from 'Stores/index';
+// import { ScrollParams } from 'react-virtualized';
```
Comment (mahdiyeh-deriv) [inline packages/reports/src/Containers/profit-table.tsx]: please remove this comment

File: packages/reports/src/Containers/profit-table.tsx [hunk 10]
```diff
@@ -1,31 +1,58 @@
...
+    return getSupportedContracts()[contract_type as TContractType] &&
+        !isForwardStarting(row_obj.shortcode, +row_obj.purchase_time_unix)
+        ? getContractPath(+row_obj.contract_id)
         : {
               component: (
                   <Localize
                       i18n_default_text="The {{trade_type_name}} contract details aren't currently available. We're working on making them available soon."
                       values={{
-                          trade_type_name: getUnsupportedContracts()[contract_type]?.name,
+                          trade_type_name: getUnsupportedContracts()[contract_type as TContractType].name,
```
Comment (mahdiyeh-deriv) [inline packages/reports/src/Containers/profit-table.tsx]: why did you remove this optional chaining operator?

File: packages/reports/src/Components/profit_loss_cell.tsx [hunk 11]
```diff
@@ -1,16 +1,15 @@
-import React from 'react';
-import PropTypes from 'prop-types';
+import React, { ReactNode } from 'react';
```
Comment (mahdiyeh-deriv) [inline packages/reports/src/Components/profit_loss_cell.tsx:1]: do we need this `{ReactNode}`?

Comment (mahdiyeh-deriv) [inline packages/reports/src/Components/profit_loss_cell.tsx]: please do not import ReactNode directly and instead use React.ReactNode

File: packages/reports/src/Containers/profit-table.tsx (hunk 10 above)
Comment (rostislav-deriv) [reply packages/reports/src/Containers/profit-table.tsx]: Not sure to be honest, added it back

**Review by mahdiyeh-deriv — APPROVED** [review]

**Review by suisin-deriv — APPROVED** [review]
